### ISSUE TITLE
[bug] tried to change option to X but it was already assigned to Y

### ISSUE THREAD

**Issue by spoofedex** [conan-io/conan#7536]
Hi all,

We are migrating some existing, legacy codebase to Conan. It is a complex network of dozens of dependencies; at one point, we have a shared library (with all requirements private), and a user of that shared library. The user and shared library both have the same dependency, with different options (which is fine, with the shared library requiring it privately).

Normally, this works fine. However, in some dependency graphs, it does fail. In the reproduction case (see below), Conan gives the following error:
```
ERROR: conflict/1.0.0@test/test: shared/1.0.0@test/test tried to change conflict/1.0.0@test/test option twice:diff to 2
but it was already assigned to 1 by change/1.0.0@test/test
```

Any ideas for workarounds are welcome, because I don't have any at the moment!

Thanks in advance!

### Environment Details
  * Operating System+version: Windows 10
  * Compiler+version: Not applicable
  * Conan version: 1.27/1.28.1
  * Python version: 3.7.3

### Steps to reproduce
Consider the following recipes:
- twice/conanfile.py:
```
from conans import ConanFile

class TwiceConan(ConanFile):
    name = "twice"
    version = "1.0.0"
    
    options = {"diff": [0, 1, 2]}
    default_options = {"diff": 0}
```
- conflict/conanfile.py:
```
from conans import ConanFile

class ConflictConan(ConanFile):
    name = "conflict"
    version = "1.0.0"
    
    requires = ["twice/1.0.0@test/test"]
```
- change/conanfile.py:
```
from conans import ConanFile

class ChangeConan(ConanFile):
    name = "change"
    version = "1.0.0"
    
    requires = ["twice/1.0.0@test/test", "conflict/1.0.0@test/test"]
    
    def configure(self):
        self.options["twice"].diff = 1
```
- shared/conanfile.py:
```
from conans import ConanFile

class SharedConan(ConanFile):
    name = "shared"
    version = "1.0.0"
    
    requires = [("change/1.0.0@test/test", "private"), ("conflict/1.0.0@test/test", "private")]
```
- user/conanfile.py:
```
from conans import ConanFile

class UserConan(ConanFile):
    name = "user"
    version = "1.0.0"
    
    requires = ["twice/1.0.0@test/test", "shared/1.0.0@test/test"]
        
    def configure(self):
        self.options["twice"].diff = 2
```
Execute the following:
```
conan export twice test/test
conan export conflict test/test
conan export change test/test
conan export shared test/test
conan create user test/test --build missing
```
This results in the error given above (see logs for more output).

Note: removing the dependency in "conflict" causes the error not to trigger.

Note: removing the configure() method from "user/conanfile.py" also causes the error not to trigger. At this point, the dependency graph is as follows (added for clarity):
![graph](https://user-images.githubusercontent.com/7325119/89910476-9401c780-dbf0-11ea-8861-43cb7ebb2b06.png)

### Logs
Output of the final "conan create":
```
[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'url'. It is recommended to add it as attribute
[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'license'. It is recommended to add it as attribute
[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'description'. It is recommended to add it as attribute
Exporting package recipe
user/1.0.0@test/test: A new conanfile.py version was exported
user/1.0.0@test/test: Folder: C:\Users\spoofedex\.conan\data\user\1.0.0\test\test\export
user/1.0.0@test/test: Using the exported files summary hash as the recipe revision: 70a76d7b9de43a153726fc89dc9c97c4
user/1.0.0@test/test: Exported revision: 70a76d7b9de43a153726fc89dc9c97c4
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=Visual Studio
compiler.runtime=MDd
compiler.version=14
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

ERROR: conflict/1.0.0@test/test: shared/1.0.0@test/test tried to change conflict/1.0.0@test/test option twice:diff to 2
but it was already assigned to 1 by change/1.0.0@test/test
```

**Comment by spoofedex** [comment]
As a temporary workaround: it _seems_ to work to add the following to the "shared" conanfile:
```
    def configure(self):
        self.options["twice"].diff = 1
```

I don't think this is desirable, and I'm not sure if it will work in all scenario's where this issue occurs. But hopefully it will allow me to proceed for now.

**Comment by jgsogo** [comment]
Hi! This is due to the graph model in Conan v1.0, there is only one instance of the `conflict` package inside the graph, only one packageID and it cannot be required with different option values. The only solution right now is to override the option value downstream in a consumer to fix the conflict (it is exactly what you have done).

For Conan 2.0 we are working in a more powerful graph and the plan is to let the user isolate some subgraphs behind the `private` keyword so the same package can appear more than once in the graph with different options. Depending on how the graph is built and requirements are consumed it can lead to ODR violations or DLL hell problems... but probably those would be the user responsibility. 

I'm sorry this is not something we can fix before Conan 2.0 (we are starting to work on it), modifying the graph model for 1.x would require too much work (and 100% we will break something).

Thanks for your comprehension!

**Comment by memsharded** [comment]
Conan 2.0 has completely changed how options are propagated. In 2.0 they will not raise conflict errors anymore. The first value assigned is the prioritized ones. Also, downstream (consumers, or command line or profiles) always have higher priority. Closing as solved.
